### PR TITLE
Add support for module namespaces

### PR DESCRIPTION
**Pull request by plbossart**
As suggested by Greg KH during the SoundWire reviews, it's not really desired to expose tons of symbols when they are intended to be used in restricted cases.

I started from the sof/intel directory and added a namespace for each existing EXPORT_SYMBOL, and conversely added a IMPORT_NS where those symbols are used.

This will  prove extremely useful when we split the clients, so make sure we don't have symbols shared by accident.

@dbaluta I didn't touch the i.mx stuff but I noticed that the sof-of-dev.c code does not use the xtensa_arch_ops variable, is this intended?

kernel documentation for reference: https://www.kernel.org/doc/html/latest/core-api/symbol-namespaces.html

### CODE REVIEW

**Comment by dbaluta** [comment]
@plbossart looks good to me. Will later send some patches to add namespace for IMX. Regarding xtensa_arch_ops at that moment I didn't understand what it's used for. Will send a patch for that too.

**Comment by dbaluta** [comment]
@plbossart OK so it seems that to add register dump support in case of problems with the DSP there is some non-trivial work to be done. I am adding this on my TODO list. Maybe late this week.

Anyhow, to catch the wave with your patches I will add support for namespace and xtensa_arch_ops today. 

One question: Why is SND_SOC_SOF_XTENSA selected from intel/Kconfig since the sof_xtensa_arch_ops is used in sof-acpi-dev.c and sof-pci-dev.c

A natural way for me would be that SND_SOC_SOF_PCI/SND_SOC_SOF_ACPI should select SND_SOC_SOF_XTENSA

Cc: @paulstelian97 this is related to your work on enhancing the debugging support for DSP.

**Comment by plbossart** [comment]
> I'm wondering whether it would be better to be less fine-graned and maybe only define a single namespace for SOF or maybe one for SOF core and one for platforms, but it looks like this finer-grained approach works well

I would like to use namespaces to detect dependencies between modules and make sure we can't have any shortcuts or functionality in the 'wrong' module. We have too many configuration options and that provides opportunities to bypass abstractions between layers.

**Comment by plbossart** [comment]
> You live and learn, had not been aware of export namespaces before. This is the first usage in sound/ as well, so probably worth giving Mark and Takashi a chance to comment on whether they want these to their tree or not.
> 
> I guess this is ok. A bit concerned with this much granularity inside SOF, how we can maintain a sane set of namespaces and to ensure correct assignment of symbols in longer terms. OTOH, this only applies to specific cases of exports, which anyways need attention, so maybe this is for the best.

It's not like I have been very creative with names, I used the Kconfigs without the CONFIG_ prefix...
I think once we start looking at multi-client or even in sof-utils.c we are going to see some files as dumping ground for stuff without any rationale in the partitioning.

**Comment by plbossart** [comment]
> One question: Why is SND_SOC_SOF_XTENSA selected from intel/Kconfig since the sof_xtensa_arch_ops is used in sof-acpi-dev.c and sof-pci-dev.c
> 
> A natural way for me would be that SND_SOC_SOF_PCI/SND_SOC_SOF_ACPI should select SND_SOC_SOF_XTENSA

Well, that's exactly what I expected to pop-up as a result of using namespaces, we have an explicit description of dependencies that are missed if you just look at modules/Kconfig.

And clearly this is wrong to me to use sof_xtensa_arch_ops in sof-acpi-dev.c or sof-pci-dev.c, this should be part of the machine description.

**Comment by plbossart** [comment]
@dbaluta see update, I moved the .arch_ops under .ops so that we can include the Xtensa namespace only at the Intel level. That should have no impact on DT/OF machines but I haven't really checked so feedback welcome

**Comment by ranj063** [comment]
@plbossart Thanks for doing this. This is super nice. When working on the multi-client stuff I was thinking along similar lines as well. There really is no proper partitioning for the core parts without the use of namespaces.

**Comment by plbossart** [comment]
> @plbossart Thanks for doing this. This is super nice. When working on the multi-client stuff I was thinking along similar lines as well. There really is no proper partitioning for the core parts without the use of namespaces.

we can thank @gregkh for mentioning the existence of those namespaces, I didn't even know they existed :-)

File: sound/soc/sof/intel/bdw.c [hunk 1]
```diff
@@ -633,13 +633,14 @@ const struct snd_sof_dsp_ops sof_bdw_ops = {
 			SNDRV_PCM_INFO_PAUSE |
 			SNDRV_PCM_INFO_BATCH,
 };
-EXPORT_SYMBOL(sof_bdw_ops);
+EXPORT_SYMBOL_NS(sof_bdw_ops, SND_SOC_SOF_BROADWELL);
+
```
Comment (dbaluta) [inline sound/soc/sof/intel/bdw.c]: Double new lines here? checkpatch should have complained.

Comment (plbossart) [reply sound/soc/sof/intel/bdw.c]: yes, fixed now, thanks!

File: sound/soc/sof/Kconfig [hunk 2]
```diff
@@ -28,7 +28,7 @@ config SND_SOC_SOF_ACPI
 	select IOSF_MBI if X86 && PCI
 	help
 	  This adds support for ACPI enumeration. This option is required
-	  to enable Intel Haswell/Broadwell/Baytrail/Cherrytrail devices
```
Comment (ranj063) [inline sound/soc/sof/Kconfig:31]: @plbossart I have a ThinkPad X1 Carbon (about 5 years old) which has a HSW CPU 
https://ark.intel.com/content/www/us/en/ark/products/76308/intel-core-i5-4300u-processor-3m-cache-up-to-2-90-ghz.html

Comment (plbossart) [reply sound/soc/sof/Kconfig:31]: I meant there are no Haswell devices with an I2S interface enabled, HDaudio is used in all commercial devices - and that prevents the use of the DSP.

**Comment by plbossart** [comment]
merged upstream

**Pull request closed without merging**